src/comm.cc

```diff
@@ -780,36 +781,41 @@ commCallCloseHandlers(int fd)
     }
 }
 
+/// sets SO_LINGER socket(7) option
+/// \param enabled -- whether linger will be active (sets linger::l_onoff)
+/// \param timeoutInSeconds -- how long to linger for (sets linger::l_linger)
+static void
+commConfigureLinger(const int fd, const OnOff enabled, const int timeoutInSeconds = 0)
+{
+    struct linger l;
+    l.l_onoff = (enabled == OnOff::on ? 1 : 0);
+    l.l_linger = timeoutInSeconds;
+
+    if (setsockopt(fd, SOL_SOCKET, SO_LINGER, reinterpret_cast<char*>(&l), sizeof(l)) < 0) {
+        const auto xerrno = errno;
+        debugs(50, DBG_CRITICAL, "ERROR: Failed to set closure behavior (SO_LINGER) for FD " << fd << ": " << xstrerr(xerrno));
+    }
+}
+
 /**
  * enable linger with time of 0 so that when the socket is
  * closed, TCP generates a RESET
  */
 void
 comm_reset_close(const Comm::ConnectionPointer &conn)
 {
-    struct linger L;
-    L.l_onoff = 1;
-    L.l_linger = 0;
-
-    if (setsockopt(conn->fd, SOL_SOCKET, SO_LINGER, (char *) &L, sizeof(L)) < 0) {
-        int xerrno = errno;
-        debugs(50, DBG_CRITICAL, "ERROR: Closing " << conn << " with TCP RST: " << xstrerr(xerrno));
+    if (Comm::IsConnOpen(conn)) {
+        commConfigureLinger(conn->fd, OnOff::on, 0);
+        debugs(5, 7, conn->id);
+        conn->close();
     }
-    conn->close();
 }
 
 // Legacy close function.
 void
 old_comm_reset_close(int fd)
 {
-    struct linger L;
-    L.l_onoff = 1;
-    L.l_linger = 0;
-
-    if (setsockopt(fd, SOL_SOCKET, SO_LINGER, (char *) &L, sizeof(L)) < 0) {
-        int xerrno = errno;
-        debugs(50, DBG_CRITICAL, "ERROR: Closing FD " << fd << " with TCP RST: " << xstrerr(xerrno));
-    }
+    commConfigureLinger(fd, OnOff::on, 0);
     comm_close(fd);
```

Comment:
If we are checking in comm_reset_close(), then we should be checking here as well, right?

```suggestion
    if (fd >= 0) {
        commConfigureLinger(fd, OnOff::on, 0);
        comm_close(fd);
    }
```

If you apply this or a similar change, please adjust PR description to cover both API methods.

Reply:
Ok, let's have these checks for symmetry. There should be no fd<0 situations now (the only caller commCloseAllSockets() iterates over positive fds).